src/main/java/redis/clients/jedis/DefaultJedisSocketConfig.java

```diff
@@ -0,0 +1,165 @@
+    return soTimeout;
+  }
+
+  @Override
+  public boolean isSsl() {
+    return ssl;
+  }
+
+  @Override
+  public SSLSocketFactory getSslSocketFactory() {
+    return sslSocketFactory;
+  }
+
+  @Override
+  public SSLParameters getSslParameters() {
+    return sslParameters;
+  }
+
+  @Override
+  public HostnameVerifier getHostnameVerifier() {
+    return hostnameVerifier;
+  }
+
+  @Override
+  public HostAndPortMapper getHostAndPortMapper() {
+    return hostAndPortMapper;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+
+    private int connectionTimeout = Protocol.DEFAULT_TIMEOUT;
+    private int soTimeout = Protocol.DEFAULT_TIMEOUT;
+
+    private boolean ssl = false;
+    private SSLSocketFactory sslSocketFactory = null;
+    private SSLParameters sslParameters = null;
+    private HostnameVerifier hostnameVerifier = null;
+
+    private HostAndPortMapper hostAndPortMapper = null;
+
+    private Builder() {
+    }
+
+    public DefaultJedisSocketConfig build() {
+      return new DefaultJedisSocketConfig(connectionTimeout, soTimeout, ssl,
+          sslSocketFactory, sslParameters, hostnameVerifier, hostAndPortMapper);
+    }
+
+    public Builder withTimeout(int timeout) {
+      return withConnectionTimeout(timeout).withSoTimeout(timeout);
+    }
+
+    public Builder withConnectionTimeout(int connectionTimeout) {
+      this.connectionTimeout = connectionTimeout;
+      return this;
+    }
+
+    public Builder withSoTimeout(int soTimeout) {
+      this.soTimeout = soTimeout;
+      return this;
+    }
+
+    public Builder withSsl(boolean ssl) {
+      this.ssl = ssl;
+      return this;
+    }
+
+    public Builder withSslSocketFactory(SSLSocketFactory sslSocketFactory) {
+      this.sslSocketFactory = sslSocketFactory;
+      return this;
+    }
+
+    public Builder withSslParameters(SSLParameters sslParameters) {
+      this.sslParameters = sslParameters;
+      return this;
+    }
+
+    public Builder withHostnameVerifier(HostnameVerifier hostnameVerifier) {
+      this.hostnameVerifier = hostnameVerifier;
+      return this;
+    }
+
+    public Builder withHostAndPortMapper(HostAndPortMapper hostAndPortMapper) {
+      this.hostAndPortMapper = hostAndPortMapper;
+      return this;
+    }
+
+    public int getConnectionTimeout() {
+      return connectionTimeout;
+    }
+
+    public int getSoTimeout() {
+      return soTimeout;
+    }
+
+    public boolean isSsl() {
+      return ssl;
+    }
+
+    public SSLSocketFactory getSslSocketFactory() {
+      return sslSocketFactory;
+    }
+
+    public SSLParameters getSslParameters() {
+      return sslParameters;
+    }
+
+    public HostnameVerifier getHostnameVerifier() {
+      return hostnameVerifier;
+    }
+
+    public HostAndPortMapper getHostAndPortMapper() {
+      return hostAndPortMapper;
+    }
+  }
+
+  static DefaultJedisSocketConfig withSsl(boolean ssl, JedisSocketConfig copy) {
```

Comment:
> I think this method signature is a bit confusing, `withSsl` implies that it's setting SSL to true, but it actually takes that as an option, can we split into withSsl and withoutSsl or something?

1. What about `public Builder withSsl(boolean ssl)` in the Builder class?
2. How about name `copyWithSsl` with same implementation?

> Should this return `JedisSocketConfig`?

No. Because
1. This is `DefaultJedisSocketConfig`, so it makes more relevant.
2. Returned DefaultJedisSocketConfig can be easily type-casted to JedisSocketConfig and set to a JedisSocketConfig variable. The opposite is not true.

Reply:
> 1. What about public Builder withSsl(boolean ssl) in the Builder class?
> 2. How about name copyWithSsl with same implementation?

Both make sense to me :)

> No. Because
>
> 1. This is DefaultJedisSocketConfig, so it makes more relevant.
> 2. Returned DefaultJedisSocketConfig can be easily type-casted to JedisSocketConfig and set to a JedisSocketConfig variable. The opposite is not true.

That sounds fair